### ISSUE TITLE
4D attention masks not compatible with generate

### ISSUE THREAD

**Issue by alex-hh** [huggingface/transformers#33549]
### System Info

Latest transformers

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

Call generate with a 4d attention mask on a llama model. Currently prepare_inputs_for_generation assumes a 2d mask (though does not check the shape, leading to a hard-to-understand error).

```
    from transformers import LlamaForCausalLM, LlamaConfig
    config = LlamaConfig(hidden_dim=64, num_attention_heads=4, num_hidden_layers=1, vocab_size=20, intermediate_size=128)
    model = LlamaForCausalLM(config)

    model.eval()
    input_ids = torch.ones(1, 10).long()
    attention_mask = torch.zeros((
        1, 1, 10, 10
    ))
    # attention_mask = torch.ones(1, 10)
    model.generate(
        input_ids=input_ids,
        max_new_tokens=2,
        do_sample=True,
        num_return_sequences=2,
        attention_mask=attention_mask,
    )
```

### Expected behavior

Generate should probably raise an informative error when passing a 4d attention mask, telling the user they should implement the required logic in e.g. _update_model_kwargs_for_generate or prepare_inputs_for_generation.

currently the errors are hard to understand, and the lack of support for this is unexpected.

**Comment by LysandreJik** [comment]
cc @gante as well

**Comment by gante** [comment]
Hi @alex-hh 👋 

Good point, we don't check that case in `generate`! Added a PR to throw an informative exception -- thank you for flagging 🤗
